---
title: "mtUMAP"
author: "Shea Andrews"
date: "6/3/2020"
output:
  html_document:
    df_print: paged
    toc: true
    number_sections: true
    toc_depth: 3
    toc_float:
      collapsed: false
      smooth_scroll: false
editor_options: 
  chunk_output_type: console
  
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)

library(tidyverse)
# library(umapr)
# library(umap)
library(FactoMineR)
library(factoextra)
library(plotly)
library(HiMC)
library(pals)
library(glue)
`%nin%` = Negate(`%in%`)

## Reading in MAP/PED files
generate_snp_data_fixed <- function (map_file, ped_file) 
{
  map <- read.csv(map_file, sep = "\t", header = FALSE, stringsAsFactors = FALSE)
  header_row <- c("Family", "Individual", "Father", "Mother", 
                  "Sex", "Phenotype")
  snps = map[, 4]
  new_header = c(header_row, snps)
  ped <- read.csv(ped_file, sep = " ", header = FALSE, stringsAsFactors = FALSE, colClasses = 'character')
  range1 = seq(1, 6, by = 1)
  snp_data = data.frame(seq(1, nrow(ped), by = 1))
  for (i in range1) {
    snp_data[, i] = ped[, i]
  }
  range2 = seq(7, ncol(ped), by = 2)
  for (i in range2) {
    index = ((i - 7)/2) + 1
    snp_data[, index + 6] = paste(ped[, i], ped[, i + 1])
  }
  names(snp_data) <- new_header
  return(snp_data)
}


```

## Dimensionality Reduction 

### PCA 
**PCA**: an orthogonal linear transformation which projects the genotype data into the new reduced dimensional space such that the greater variance comes in an order

### UMAP 
**UMAP**: a novel non-linear dimensionality reduction technique based on Riemannian geometry and algebraic topology to model and preserve the high-dimensional topology of data points in the lowdimensional space

**PCA–UMAP**: an application of UMAP for principal components of genotype data to be computationally more advantageous and statistically less noisy

UMAP has several hyperparameters that can have a signficant impact on the resulting embedding. 

+ `n_neighbors`: The size of local neighborhood (in terms of number of neighboring sample points) used for manifold approximation. Larger values result in more global views of the manifold, while smaller values result in more local data being preserved. In general values should be in the range 2 to 100 (default = 15).
  - Using few neighbours to build a manifold emphasizes closer relatedness, with related samples clustered togeather.
+ `min_dist`: The effective minimum distance between embedded points. Smaller values will result in a more clustered/clumped embedding where nearby points on the manifold are drawn closer together, while larger values will result on a more even dispersal of points. The value should be set relative to the “spread“ value. (default = 0.1) 
  - spread can therefore be used to control the inter-cluster distances to some extent, where as min_dist controls the size of the clusters.
+ `n_components`: The dimension of the space to embed into (defaults = 2)
+ `metric`: The metric to use to compute distances in high dimensional space (default = euclidean)
+ `n pca`: For genetic ancestry the number of PCs used in the data input. Adding more components results in progressively finer clusters until approximately 20 populations appear using 15 components; adding further components converges to results similar to using the entire genotype data

### References 

#### Papers 
+ Diaz-Papkovich, A., et al. (2019). UMAP reveals cryptic population structure and phenotype heterogeneity in large genomic cohorts. [PLOS Genetics  15(11), e1008432](https://dx.doi.org/10.1371/journal.pgen.1008432)
  - data = 1kg, UKBB, HRS; metric = Eucledian, PCs = 10; n_neigbours = 15, min_dist = 0.5
  - Parameter Tuning
    * PCs: 2 - 15  
    * min_dist:  0.001 - 0.5 
    * n_neigbours: 5 - 50
+ Karczewski, K., et al. (2020). The mutational constraint spectrum quantified from variation in 141,456 humans. [Nature  581(7809), 434-443](https://dx.doi.org/10.1038/s41586-020-2308-7)
  - data = gnomAD; PCs = 7; min_dist = 0.5; n_neigbours = 30
+ Sakaue, S., et al. (2020). Dimensionality reduction reveals fine-scale structure in the Japanese population with consequences for polygenic risk prediction. [Nature Communications  11(1), 1569](https://dx.doi.org/10.1038/s41467-020-15194-z)
  - data = Japanse Biobank; PCs = 50; min_dist = 0.1; n_neigbours = 15; 
+ Yamamoto, K.,et al. (2020). Genetic and phenotypic landscape of the mitochondrial genome in the Japanese population. [Communications Biology  3(1), 104](https://dx.doi.org/10.1038/s42003-020-0812-9)
  - data = Japanese pop; PCs = 20; n_neighbors = 100; min_dist = 0.99
  
#### Documentation 
+ [umapr](https://github.com/ropenscilabs/umapr): wraps the Python implementation of UMAP to make the algorithm accessible from within R
+ [umap](https://github.com/tkonopka/umap): R package that provides an interface to the UMAP algorithm, including a translation of the original algorithm into R with minimal dependencies
  - [Uniform Manifold Approximation and Projection in R](https://cran.r-project.org/web/packages/umap/vignettes/umap.html) 
+ [uwot](https://github.com/jlmelville/uwot): R package implementing the UMAP dimensionality reduction method that also implements the supervised and metric (out-of-sample) learning extensions to the basic method.
+ [umap](https://umap-learn.readthedocs.io/en/latest/reproducibility.html): Read the docs for the python implementation

```{r default_param}
param_umapr <- tribble(
  ~parameter, ~umapr,
  "include_input",  "TRUE", 
  "n_neighbors",  "15",
  "n_components",  "2", 
  "metric",  "euclidean", 
  "n_epochs",  "NULL",
  "learning_rate",  "1", 
  "alpha",  "1", 
  "init",  "spectral", 
  "spread",  "1",
  "min_dist",  "0.1", 
  "set_op_mix_ratio",  "1", 
  "local_connectivity",  "1",
  "repulsion_strength",  "1", 
  "bandwidth",  "1", 
  "gamma",  "1",
  "negative_sample_rate",  "5", 
  "transform_queue_size",  "4", 
  "a",  "NULL",
  "b",  "NULL", 
  "random_state",  "NULL",
  "metric_kwds",  "dict()",
  "angular_rp_forest",  "FALSE", 
  "target_n_neighbors",  "-1",
  "target_metric",  "categorical",
  "target_metric_kwds",  "dict()",
  "target_weight",  "0.5", 
  "transform_seed",  "42")


param_umap <- tribble(
  ~parameter, ~umap,
  "n_neighbors",  "15",
  "n_components",  "2",
  "metric",  "euclidean",
  "n_epochs",  "200",
  "input",  "data",
  "init",  "spectral",
  "min_dist",  "0.1",
  "set_op_mix_ratio",  "1",
  "local_connectivity",  "1",
  "bandwidth",  "1",
  "alpha",  "1",
  "gamma",  "1",
  "negative_sample_rate",  "5",
  "a",  "NULL",
  "b",  "NULL",
  "spread",  "1",
  "random_state",  "NULL",
  "transform_state",  "NULL",
  "knn_repeats",  "1"

)

param_uwot <- tribble(
  ~parameter, ~uwot,
  "n_neighbors", "15",
  "n_components", "2",
  "metric", "euclidean",
  "n_epochs", "NULL",
  "learning_rate", "1",
  "scale", "FALSE",
  "init", "spectral",
  "init_sdev", "NULL",
  "spread", "1",
  "min_dist", "0.01",
  "set_op_mix_ratio", "1",
  "local_connectivity", "1",
  "bandwidth", "1",
  "repulsion_strength", "1",
  "negative_sample_rate", "5",
  "a", "NULL",
  "b", "NULL",
  "nn_method", "NULL",
  "n_trees", "50",
  "search_k", "2 * n_neighbors * n_trees",
  "approx_pow", "FALSE",
  "y", "NULL",
  "target_n_neighbors", "n_neighbors",
  "target_metric", "euclidean",
  "target_weight", "0.5",
  "pca", "NULL",
  "pca_center", "TRUE",
  "pcg_rand", "TRUE",
  "fast_sgd", "FALSE",
  "ret_model", "FALSE",
  "ret_nn", "FALSE",
  "ret_extra", "c()",
  "n_threads", "NULL",
  "n_sgd_threads", "0",
  "grain_size", "1"
)

param_umapr %>% left_join(param_umap) %>% left_join(param_uwot) %>% 
  knitr::kable() %>%
  kableExtra::kable_styling()

```


## mtReference Panel 

+ Read in map/ped files 
+ assign mitochondrial haplogroups using Hi-Mc
+ Convert alleles to 0/1 coding 

```{r mtreference-wrangle, cache=T}
## Provenence 
country <- read_tsv("/Users/sheaandrews/GitCode/MitoImputePrep/DerivedData/ReferencePanel_v6/ReferencePanel_v1_highQual_MAF0.01_filtered_countryInfo.ped") %>% 
  select(Family, Individual, geographic_data, Continent, Country, Region)

## Haplogroup assignments 
dat <- generate_snp_data_fixed("/Users/sheaandrews/GitCode/MitoImpute/ReferencePanel/ReferencePanel.map",
                               "/Users/sheaandrews/GitCode/MitoImpute/ReferencePanel/ReferencePanel.ped")

data(nodes)
classifications.raw <- HiMC::getClassifications(dat)
classifications <- as_tibble(classifications.raw[-nrow(classifications.raw),]) #remove Mitochondria_Information row 

ped <- dat %>% 
  magrittr::set_colnames(
    c("Family", "Individual", "Father", "Mother", "Sex", "Phenotype", 
      paste0('mt', colnames(magrittr::extract(., 7:ncol(.))))
      )
    ) %>% 
  as_tibble() %>%
  na_if(., "0 0") %>% 
  mutate_at(vars(starts_with("mt")), as_factor) %>% 
  mutate_at(vars(starts_with("mt")), as.numeric) %>% 
  left_join(classifications) %>% 
  left_join(country) %>% 
  mutate(macro = ifelse(str_detect(haplogroup, "^L"),
                        substr(haplogroup, start = 1, stop = 2),
                        substr(haplogroup, start = 1, stop = 1)), 
         Continent = replace_na(Continent, "Unknown")) %>% 
  select(Family, Individual, Father, Mother, Sex, Phenotype, haplogroup, macro, 
         geographic_data, Continent, Country, Region, everything()) %>% 
  filter(macro %nin% c("u")) %>% 
  filter(!is.na(macro))

```

```{r, eval=FALSE}
count(ped, Continent)
```

### PCA analysis

Run PCA analysis of mtReference Panel 

+ extract results for individules
+ plot 2D and 3D results

```{r mtref-pca}
## PCA 
res.pca <- ped %>% 
  select(starts_with("mt")) %>% 
  PCA(., scale.unit = F, ncp = 10, graph = FALSE)

ind.pca <- get_pca_ind(res.pca)$coord %>% 
  as_tibble() %>%
  bind_cols(select(ped, Individual, haplogroup, macro, Continent)) %>% 
  select(Individual, haplogroup, macro, Continent, everything()) 
```

```{r}
eig.val <- get_eigenvalue(res.pca)
fviz_eig(res.pca, addlabels = TRUE, ylim = c(0, 50))
```

```{r mtref-plot-2d-pca}

ggplot(ind.pca, aes(x = Dim.1, y = Dim.2, colour = macro)) + 
  scale_color_manual(values = cols25()) + 
  geom_point() + theme_bw()

ggplot() + 
  geom_point(data = filter(ind.pca, Continent == "Unknown"), aes(x = Dim.1, y = Dim.2, colour = Continent)) + 
  geom_point(data = filter(ind.pca, Continent != "Unknown"), aes(x = Dim.1, y = Dim.2, colour = Continent)) + 
  scale_color_manual(values = cols25()) + 
  theme_bw()
```



```{r mtref-pca-plot-3d}
plot_ly(ind.pca, x = ~Dim.2, y = ~Dim.1, z = ~Dim.3, color = ~macro, colors = cols25(),
        type = 'scatter3d', mode = 'markers', marker = list(size = 3), 
        hoverinfo = 'text', text = ~paste('</br> ID: ', Individual, 
                                          '</br> Haplogroup: ', macro,
                                          '</br> PC1: ', round(Dim.1, 2),
                                          '</br> PC2: ', round(Dim.2, 2),
                                          '</br> PC3: ', round(Dim.3, 2)))

plot_ly(ind.pca, x = ~Dim.2, y = ~Dim.1, z = ~Dim.3, color = ~Continent, colors = cols25(),
        type = 'scatter3d', mode = 'markers', marker = list(size = 3), 
        hoverinfo = 'text', text = ~paste('</br> ID: ', Individual, 
                                          '</br> Haplogroup: ', macro,
                                          '</br> PC1: ', round(Dim.1, 2),
                                          '</br> PC2: ', round(Dim.2, 2),
                                          '</br> PC3: ', round(Dim.3, 2)))

```

### UMAP 

```{r mtref-umap}
embedding_mtref_2d <- select(ind.pca, starts_with("Dim")) %>% 
  umapr::umap(., min_dist = 0.5, n_neighbor = 100, n_components = 2)  %>% as.tibble() %>%
  bind_cols(select(ped, Individual, haplogroup, macro, Continent)) 

ggpubr::ggarrange(
  ggplot(embedding_mtref_2d, aes(x = UMAP1, y = UMAP2, color = macro)) + 
  geom_point() + 
  scale_color_manual(values = cols25()) +
  theme_bw(),

ggplot() + 
  geom_point(data = filter(embedding_mtref_2d, Continent == "Unknown"), aes(x = UMAP1, y = UMAP2, colour = Continent)) + 
  geom_point(data = filter(embedding_mtref_2d, Continent != "Unknown"), aes(x = UMAP1, y = UMAP2, colour = Continent)) + 
  scale_color_manual(values = cols25()) +
  theme_bw()
)

embedding_mtref_3d <- select(ind.pca, starts_with("Dim")) %>% 
  umapr::umap(., min_dist = 0.5, n_neighbor = 100, n_components = 3)  %>% as.tibble() %>%
  bind_cols(select(ped, Individual, haplogroup, macro, Continent)) 

plot_ly(embedding_mtref_3d, x = ~UMAP2, y = ~UMAP1, z = ~UMAP3, color = ~macro, colors = cols25(), type = 'scatter3d', mode = 'markers', marker = list(size = 3), 
        hoverinfo = 'text', text = ~paste('</br> ID: ', Individual, 
                                          '</br> Haplogroup: ', macro,
                                          '</br> PC1: ', round(UMAP1, 2),
                                          '</br> PC2: ', round(UMAP2, 2),
                                          '</br> PC3: ', round(UMAP3, 2)))

plot_ly(filter(embedding_mtref_3d, Continent != "Unknown"), x = ~UMAP2, y = ~UMAP1, z = ~UMAP3, color = ~Continent, colors = cols25(), type = 'scatter3d', mode = 'markers', marker = list(size = 3), 
        hoverinfo = 'text', text = ~paste('</br> ID: ', Individual, 
                                          '</br> Haplogroup: ', macro,
                                          '</br> PC1: ', round(UMAP1, 2),
                                          '</br> PC2: ', round(UMAP2, 2),
                                          '</br> PC3: ', round(UMAP3, 2)))

```


```{r, mtref}
min_dist = 0.5
n_neighbor = 100
n_components = 2
n_epochs = 2000
init = "random"

mtref_umapr_tab <- select(ind.pca, starts_with("Dim")) %>% 
  umapr::umap(., min_dist = min_dist, n_neighbor = n_neighbor, n_components = n_components, n_epochs = NULL, init = init) %>%
  as.tibble() %>%
  bind_cols(select(ped, Individual, haplogroup, macro, Continent))

p1 <- ggplot(mtref_umapr_tab, aes(x = UMAP1, y = UMAP2, color = macro)) + 
  geom_point() + 
    scale_color_manual(values = cols25()) +
    theme_bw() + 
  labs(title = "UMAPR")

mtref_umap_res <- select(ind.pca, starts_with("Dim")) %>% 
  umap::umap(., min_dist = min_dist, n_neighbor = n_neighbor, n_components = n_components, n_epochs = n_epochs, init = init)  

mtref_umap_tab <- embedding_mtref_umap_res %>% magrittr::use_series(layout) %>% 
  as.tibble() %>%
  bind_cols(select(ped, Individual, haplogroup, macro, Continent)) 

p2 <- ggplot(mtref_umap_tab, aes(x = V1, y = V2, color = macro)) + 
  geom_point() + 
    scale_color_manual(values = cols25()) +
    theme_bw() + 
  labs(title = "UMAP")

mtref_uwot_tab <- select(ind.pca, starts_with("Dim")) %>% 
  uwot::umap(., min_dist = min_dist, n_neighbor = n_neighbor, n_components = n_components, n_epochs = n_epochs, init = init) %>% 
  as.tibble() %>%
  bind_cols(select(ped, Individual, haplogroup, macro, Continent))

p3 <- ggplot(mtref_uwot_tab, aes(x = V1, y = V2, color = macro)) + 
  geom_point() + 
    scale_color_manual(values = cols25()) +
    theme_bw() + 
  labs(title = "UWOT")

ggpubr::ggarrange(
  p1, p2, p3,  ncol = 3, legend = "bottom", common.legend = TRUE
)


```


## Thousand Genomes 

### Data Wrangling
```{r 1kg-wrangle, cache=T}
## Haplogroup assignments 
dat_1kg <- generate_snp_data_fixed(
  "/Users/sheaandrews/GitCode/MitoImputePrep/DerivedData/ThousandGenomes/chrMT_1kg_norm_decomposed_firstAlt.map",
  "/Users/sheaandrews/GitCode/MitoImputePrep/DerivedData/ThousandGenomes/chrMT_1kg_norm_decomposed_firstAlt.ped")
info_1kg <- read_tsv("/Users/sheaandrews/GitCode/MitoImputePrep/data/ThousandGenomes/20130606_g1k.ped", guess_max = 10000) %>% 
  select(Individual = `Individual ID`, Family_ID = `Family ID`, Father = `Paternal ID`, Mother = `Maternal ID`, Population, Gender) %>% 
  mutate(cont = case_when(
    Population == "ESN" ~ "AFR",
    Population == "GWD" ~ "AFR",
    Population == "LWK" ~ "AFR",
    Population == "MSL" ~ "AFR",
    Population == "YRI" ~ "AFR",
    Population == "JPT" ~ "EAS",
    Population == "CHB" ~ "EAS",
    Population == "CDX" ~ "EAS",
    Population == "CHS" ~ "EAS",
    Population == "KHV" ~ "EAS",
    Population == "CEU" ~ "EUR",
    Population == "FIN" ~ "EUR",
    Population == "GBR" ~ "EUR",
    Population == "IBS" ~ "EUR",
    Population == "TSI" ~ "EUR",
    Population == "BEB" ~ "IND",
    Population == "GIH" ~ "IND",
    Population == "ITU" ~ "IND",
    Population == "PJL" ~ "IND",
    Population == "STU" ~ "IND",
    Population == "ACB" ~ "AMR",
    Population == "ASW" ~ "AMR",
    Population == "CLM" ~ "AMR",
    Population == "MXL" ~ "AMR",
    Population == "PEL" ~ "AMR",
    Population == "PUR" ~ "AMR",
  ))

data(nodes)
classifications_1kg.raw <- HiMC::getClassifications(dat_1kg)
classifications_1kg <- as_tibble(classifications_1kg.raw[-nrow(classifications_1kg.raw),]) #remove Mitochondria_Information row 

ped_1kg <- dat_1kg %>% 
  magrittr::set_colnames(
    c("Family", "Individual", "Father", "Mother", "Sex", "Phenotype", 
      paste0('mt', colnames(magrittr::extract(., 7:ncol(.))))
      )
    ) %>% 
  as_tibble() %>%
  na_if(., "0 0") %>% 
  mutate_at(vars(starts_with("mt")), as_factor) %>% 
  mutate_at(vars(starts_with("mt")), as.numeric) %>% 
  left_join(classifications_1kg) %>% 
  mutate(macro = ifelse(str_detect(haplogroup, "^L"),
                        substr(haplogroup, start = 1, stop = 2),
                        substr(haplogroup, start = 1, stop = 1))) %>% 
  select(-Family, -Sex, -Father, -Mother) %>%
  left_join(info_1kg, by = "Individual") %>%
  select(Family = Family_ID, Individual, Father, Mother, Sex = Gender, Phenotype, 
         pop = Population, haplogroup, macro, full_path, everything()) %>% 
  filter(macro %nin% c("u")) %>% 
  filter(!is.na(macro)) # %>% 
  # group_by(Family) %>% 
  # slice(1) %>% 
  # ungroup()
#  filter(macro %nin% c("L0", "L1", "L2", "L3"))

miss_gt <- ped_1kg %>%
  select(everything()) %>%  # replace to your needs
  summarise_all( ~ sum(is.na(.))) %>% 
  t() %>% 
  as_tibble(rownames = "var") %>% 
  filter(V1 > 0) %>% pull(var)
```

```{r tabs, eval=FALSE}

count(ped_1kg, pop) %>% print(n = Inf)
count(ped_1kg, cont) %>% print(n = Inf)
count(ped_1kg, macro) %>% print(n = Inf)
count(ped_1kg, haplogroup) %>% print(n = Inf)

```

### PCA 

```{r 1kg-pca}

## PCA 
res.pca_1kg <- ped_1kg %>% 
  select(starts_with("mt")) %>% 
  select(-all_of(miss_gt)) %>%
#  mutate_all(., .funs = function(x){ifelse(is.na(x), round(mean(x, na.rm = TRUE)), x)}) %>%
  PCA(., scale.unit = F, ncp = 10, graph = FALSE)

ind.pca_1kg <- get_pca_ind(res.pca_1kg)$coord %>% 
  as_tibble() %>%
  bind_cols(select(ped_1kg, Individual, haplogroup, macro, pop, cont)) %>% 
  select(Individual, haplogroup, macro, everything()) 

```

```{r 1kg-plot-scree}
eig.val_1kg <- get_eigenvalue(res.pca_1kg)
fviz_eig(res.pca_1kg, addlabels = TRUE, ylim = c(0, 50))

```


```{r 1kg-plot-2d-pca}

ggplot(ind.pca_1kg, aes(x = Dim.1, y = Dim.2, colour = macro)) + 
  geom_point() + 
  scale_color_manual(values = cols25()) + 
  theme_bw() + theme(legend.position = "bottom")

ggplot(ind.pca_1kg, aes(x = Dim.1, y = Dim.2, colour = cont)) + 
  geom_point() + 
  scale_color_manual(values = cols25()) + 
  theme_bw() + theme(legend.position = "bottom")

```

```{r 1kg-plot-3d-pca}
plot_ly(ind.pca_1kg, x = ~Dim.2, y = ~Dim.1, z = ~Dim.3, color = ~macro, colors = cols25(), 
        type = 'scatter3d', mode = 'markers', marker = list(size = 3), 
        hoverinfo = 'text', text = ~paste('</br> ID: ', Individual, 
                                          '</br> Haplogroup: ', macro,
                                          '</br> PC1: ', round(Dim.1, 2),
                                          '</br> PC2: ', round(Dim.2, 2),
                                          '</br> PC3: ', round(Dim.3, 2)))

plot_ly(ind.pca_1kg, x = ~Dim.2, y = ~Dim.1, z = ~Dim.3, color = ~cont, colors = cols25(), 
        type = 'scatter3d', mode = 'markers', marker = list(size = 3), 
        hoverinfo = 'text', text = ~paste('</br> ID: ', Individual, 
                                          '</br> Haplogroup: ', macro,
                                          '</br> PC1: ', round(Dim.1, 2),
                                          '</br> PC2: ', round(Dim.2, 2),
                                          '</br> PC3: ', round(Dim.3, 2)))

```

### UMAP 

```{r 1kg-umap}
## UMAP 
embedding_mt_2d <- ped_1kg %>% 
  select(starts_with("mt")) %>% 
  select(-all_of(miss_gt)) %>%
#  mutate_all(., .funs = function(x){ifelse(is.na(x), round(mean(x, na.rm = TRUE)), x)}) %>%
  umapr::umap(., min_dist = 0.5, n_neighbor = 50, n_components = 2) %>% 
  as.tibble() %>%
  bind_cols(select(ped_1kg, Individual, haplogroup, macro, pop, cont)) 

embedding_mt_3d <- ped_1kg %>% 
  select(starts_with("mt")) %>% 
  select(-all_of(miss_gt)) %>%
#  mutate_all(., .funs = function(x){ifelse(is.na(x), round(mean(x, na.rm = TRUE)), x)}) %>%
  umapr::umap(., min_dist = 0.5, n_neighbor = 50, n_components = 3) %>% 
  as.tibble() %>%
  bind_cols(select(ped_1kg, Individual, haplogroup, macro, pop, cont)) 

## PCA+UMAP
embedding_pca_2d <- select(ind.pca_1kg, starts_with("Dim")) %>% 
  umapr::umap(., min_dist = 0.5, n_neighbor = 50, n_components = 2) %>% 
  as.tibble() %>%
  bind_cols(select(ped_1kg, Individual, haplogroup, macro, pop, cont)) 

embedding_pca_3d <- select(ind.pca_1kg, starts_with("Dim")) %>%
  umapr::umap(., min_dist = 0.5, spread = 1, n_neighbor = 50, n_components = 3) %>% 
  as.tibble() %>%
  bind_cols(select(ped_1kg, Individual, haplogroup, macro, pop, cont)) 

```

```{r 1kg-umap-plot-2d}
ggpubr::ggarrange(
  ggplot(embedding_mt_2d, aes(x = UMAP1, y = UMAP2, color = macro)) + 
    geom_point() + scale_color_manual(values = cols25()) +  theme_bw() + labs(title = "UMAP"),
  ggplot(embedding_pca_2d, aes(x = UMAP1, y = UMAP2, color = macro)) + 
    geom_point() + scale_color_manual(values = cols25()) +  theme_bw() + labs(title = "PCA+UMAP"), 
  common.legend = TRUE, legend = "bottom"
)

ggpubr::ggarrange(
  ggplot(embedding_pca_2d, aes(x = UMAP1, y = UMAP2, color = cont)) + 
    geom_point() + scale_color_manual(values = cols25()) +  theme_bw() + labs(title = "PCA+UMAP: Continent"),
  ggplot(embedding_pca_2d, aes(x = UMAP1, y = UMAP2, color = macro)) + 
    geom_point() + scale_color_manual(values = cols25()) +  theme_bw() + labs(title = "PCA+UMAP: mtHg")
)


```

```{r 1kg-umap-plot-3d}
plot_ly(embedding_mt_3d, x = ~UMAP2, y = ~UMAP1, z = ~UMAP3, color = ~macro, colors = cols25(),
        type = 'scatter3d', mode = 'markers', marker = list(size = 3),
        hoverinfo = 'text', text = ~paste('</br> ID: ', Individual, 
                                          '</br> Macro: ', macro,
                                          '</br> Haplogroup: ', haplogroup,
                                          '</br> PC1: ', round(UMAP1, 2),
                                          '</br> PC2: ', round(UMAP2, 2),
                                          '</br> PC3: ', round(UMAP3, 2))) %>%  
        layout(title="UMAP")

plot_ly(embedding_pca_3d, x = ~UMAP2, y = ~UMAP1, z = ~UMAP3, color = ~macro, colors = cols25(), 
        type = 'scatter3d', mode = 'markers', marker = list(size = 3), 
        hoverinfo = 'text', text = ~paste('</br> ID: ', Individual, 
                                          '</br> Macro: ', macro,
                                          '</br> Haplogroup: ', haplogroup,
                                          '</br> Pop: ', pop,
                                          '</br> Continent: ', cont,
                                          '</br> PC1: ', round(UMAP1, 2),
                                          '</br> PC2: ', round(UMAP2, 2),
                                          '</br> PC3: ', round(UMAP3, 2))) %>%  
        layout(title="PCA+UMAP")

plot_ly(embedding_pca_3d, x = ~UMAP2, y = ~UMAP1, z = ~UMAP3, color = ~cont, colors = cols25(), 
        type = 'scatter3d', mode = 'markers', marker = list(size = 3), 
        hoverinfo = 'text', text = ~paste('</br> ID: ', Individual, 
                                          '</br> Macro: ', macro,
                                          '</br> Haplogroup: ', haplogroup,
                                          '</br> Pop: ', pop,
                                          '</br> Continent: ', cont,
                                          '</br> PC1: ', round(UMAP1, 2),
                                          '</br> PC2: ', round(UMAP2, 2),
                                          '</br> PC3: ', round(UMAP3, 2))) %>%  
        layout(title="PCA+UMAP")


```

#### Hypurr-ameters 

```{r hyperparameters, cache=TRUE}
### UMAP Function 
umap_helper <- function(data, pcs, min_dist, n_neighbor, n_components){
  pb$tick()$print() ## Progress bar
  out <- select(data, Dim.1:glue("Dim.", pcs)) %>% 
    umapr::umap(., min_dist = min_dist, spread = 1, n_neighbor = n_neighbor, n_components = n_components) %>% 
    as.tibble() %>%
    bind_cols(select(data, Individual, haplogroup, macro, pop, cont)) 
  out
}

### Hyper-parameters to iterative over 
umap_param <- tibble(pcs = 2:3, 
                     min_dist = rep(0.5, length(pcs)), 
                     n_neighbor = rep(50, length(pcs)), 
                     n_components = rep(2,3, length(pcs))
)

umap_param <- expand_grid(
  pcs = 2:10, 
  n_components = 2, 
  min_dist = c(0.1, 0.25, 0.5, 0.75, 0.99), 
  n_neighbor = c(15, 25, 50, 75, 100)
)

pb <- progress_estimated(nrow(umap_param)) ## initiate Progress bar
hp_1kg_embedding <- umap_param %>% 
  rowwise() %>% 
  mutate(res = list(umap_helper(ind.pca_1kg, pcs = pcs, min_dist = min_dist, 
                                n_neighbor = n_neighbor, n_components = n_components)))
```


```{r}
hp_1kg_embedding_plots <- hp_1kg_embedding %>% 
  mutate(plots = list(
    ggplot(res, aes(x = UMAP1, y = UMAP2, color = macro)) +
      geom_point() + scale_color_manual(values = cols25()) + theme_bw()
    ))

## PC Plots (n_components == 2, min_dist == 0.5, n_neighbor == 50)
ggpubr::ggarrange(plotlist =  filter(hp_1kg_embedding_plots, 
                                     pcs %in% c(2:10), 
                                     n_components == 2, 
                                     min_dist == 0.1, 
                                     n_neighbor == 15)  %>% pull(plots), 
                  legend = "bottom", 
                  common.legend = TRUE, 
                  labels = glue("PCs: {npc}", npc = 2:10))

## n_neigbor (pcs == 10, n_components == 2, min_dist == 0.5)
ggpubr::ggarrange(plotlist =  filter(hp_1kg_embedding_plots, 
                                     pcs == 10, 
                                     n_components == 2, 
                                     min_dist == 0.1, 
                                     n_neighbor %in% c(10, 25, 50, 75, 100)) %>% 
                    pull(plots), 
                  legend = "bottom", 
                  common.legend = TRUE, 
                  labels = glue("NN: {npc}", npc = c(10, 25, 50, 75, 100)))

## m_dist (pcs == 10, n_components == 2, n_neighbor == 50)
ggpubr::ggarrange(plotlist = filter(hp_1kg_embedding_plots, 
                                    pcs == 10, 
                                     n_components == 2, 
                                     min_dist %in% c(0.1, 0.25, 0.5, 0.75, 0.99), 
                                     n_neighbor == 15) %>% 
                    pull(plots), 
                  legend = "bottom", 
                  common.legend = TRUE, 
                  labels = glue("MD: {npc}", npc = c(0.1, 0.25, 0.5, 0.75, 0.99)))


```


## Projection 

```{r}
umap_pca_2d <- select(ind.pca_1kg, starts_with("Dim")) %>% 
  umap::umap(., min_dist = 0.5, n_neighbor = 50, n_components = 2) %>% 
  magrittr::extract2(1) %>%
  as.tibble() %>% 
  rename(UMAP1 = V1, UMAP2 = V2) %>%
  bind_cols(ind.pca_1kg) %>%
  select(Individual, haplogroup, macro, pop, cont, starts_with("Dim"), starts_with("UMAP"))

ggpubr::ggarrange(
  ggplot(embedding_pca_2d, aes(x = UMAP1, y = UMAP2, color = macro)) + 
    geom_point() + scale_color_manual(values = cols25()) +  theme_bw() + labs(title = "umapr::umap"),
  ggplot(umap_pca_2d, aes(x = UMAP1, y = UMAP2, color = macro)) + 
    geom_point() + scale_color_manual(values = cols25()) +  theme_bw() + labs(title = "umap::umap")
)

ped_joint <- bind_rows(
  ped %>%
    mutate(data = "mtRef") %>%
    select(Individual, haplogroup, macro, data, starts_with("mt")), 
  
  ped_1kg %>%
    mutate(data = "tkg") %>%
    select(Individual, haplogroup, macro, data, starts_with("mt"))
) %>% 
  arrange(data, Individual)


miss_gt_joint <- ped_joint %>%
  select(starts_with("mt")) %>%  # replace to your needs
  summarise_all( ~ sum(is.na(.))) %>% 
  t() %>% 
  as_tibble(rownames = "var") %>% 
  filter(V1 > 1000) %>% pull(var)

## PCA 
ped_joint_df <- ped_joint %>% 
  select(-all_of(miss_gt_joint)) %>% 
  split(.$haplogroup) %>%
  map(., ~mutate_all(., .funs = function(x){ifelse(is.na(x), round(mean(x, na.rm = TRUE)), x)})) %>%
  bind_rows() %>% 
  arrange(data, Individual) %>%
  split(.$data) %>% 
  map(., select, starts_with("mt")) 

res.pca_joint <- ped_joint_df %>%
  map(~PCA(., scale.unit = F, ncp = 10, graph = FALSE))

test <- predict(magrittr::extract2(res.pca_joint, "tkg"), magrittr::extract2(ped_joint_df, "mtRef")) %>% 
  magrittr::use_series(coord) %>% 
  as_tibble() %>% 
  bind_cols(select(filter(ped_joint, data == "mtRef"), Individual, haplogroup, macro, data)) %>% 
  mutate(data = "mtRef_pred") %>% 
  select(Individual, data, haplogroup, macro, everything()) 
    
ind.pca_joint <- res.pca_joint %>% 
  map(get_pca_ind) %>% 
  map(magrittr::use_series, coord) %>% 
  map(as_tibble) %>%
  bind_rows() %>%
  bind_cols(select(ped_joint, Individual, haplogroup, macro, data)) %>% 
  select(Individual, data, haplogroup, macro, everything()) %>% 
  bind_rows(test) %>% 
  split(.$data)

test2 <- bind_rows(test, ind.pca_1kg)

ggplot(test2, aes(x = Dim.1, y = Dim.2, colour = macro)) + geom_point() + facet_grid(. ~ data) + 
  scale_color_manual(values = cols25())


custom.config = umap::umap.defaults
custom.config$random_state = 333
custom.config$transform_state = 333
custom.config$n_neighbor = 50
custom.config$min_dist = 0.99
custom.config$spread = 2
  
## 1KG UMAP
tkg.umap_res = ind.pca_joint %>% 
 magrittr::use_series(tkg) %>% 
 # select(starts_with("Dim")) %>%
 select(Dim.1:Dim.3) %>%
  umap::umap(., custom.config)

tkg.umap_res_tab <- tkg.umap_res %>% 
  magrittr::use_series(layout) %>% 
  as_tibble() %>% 
  bind_cols(magrittr::use_series(ind.pca_joint, tkg)) %>% 
  select(Individual, data, haplogroup, macro, UMAP1 = V1, UMAP2 = V2, everything())

p1 <- ggplot(tkg.umap_res_tab, aes(x = UMAP1, y = UMAP2, colour = macro)) + geom_point() + facet_grid(. ~ data) + 
  scale_color_manual(values = cols25())
p1 

## Projection of mtRef
mtRef_pred.umap_res = predict(tkg.umap_res, 
                           magrittr::use_series(ind.pca_joint, mtRef_pred) %>% 
                             # select(starts_with("Dim"))
                             select(Dim.1:Dim.3)
                           ) %>% 
  as_tibble()  %>% 
  bind_cols(magrittr::use_series(ind.pca_joint, mtRef_pred)) %>% 
  select(Individual, data, haplogroup, macro, UMAP1 = V1, UMAP2 = V2, everything()) 

p3 <- ggplot(mtRef_pred.umap_res, aes(x = UMAP1, y = UMAP2, colour = macro)) + geom_point() + facet_grid(. ~ data) + 
  scale_color_manual(values = cols25())
p3

ggplot() + 
  geom_point(data = mtRef_pred.umap_res, aes(x = UMAP1, y = UMAP2)) + 
  geom_point(data = tkg.umap_res_tab, aes(x = UMAP1, y = UMAP2, color = macro)) + 
  scale_color_manual(values = cols25())

## mtRef UMAP
mtRef.umap_res = ind.pca_joint %>% 
 magrittr::use_series(mtRef) %>% 
  select(starts_with("Dim")) %>%
  umap::umap(.)

mtRef.umap_res_tab <- mtRef.umap_res %>% 
  magrittr::use_series(layout) %>% 
  as_tibble() %>% 
  bind_cols(magrittr::use_series(ind.pca_joint, mtRef)) %>% 
  select(Individual, data, haplogroup, macro, UMAP1 = V1, UMAP2 = V2, everything())

p2 <- ggplot(mtRef.umap_res_tab, aes(x = UMAP1, y = UMAP2, colour = macro)) + geom_point() + facet_grid(. ~ data) + 
  scale_color_manual(values = cols25())


ggpubr::ggarrange(
  p2, p3, p1,  ncol = 3, legend = "bottom", common.legend = TRUE
)



```


```{r}
iris.data = iris[, grep("Sepal|Petal", colnames(iris))]
iris.umap_res = umap::umap(iris.data)
iris.umap_df <- bind_cols(as_tibble(iris), as_tibble(iris.umap$layout))

ggplot(iris.umap_df, aes(x = V1, y = V2, color = Species)) + geom_point()


iris.wnoise = iris.data + matrix(rnorm(150*40, 0, 0.1), ncol=4)
colnames(iris.wnoise) = colnames(iris.data)
head(iris.wnoise, 3)

iris.wnoise.umap = predict(iris.umap, iris.wnoise)
iris.wnoise_umap_df <- bind_cols(as_tibble(iris.wnoise), as_tibble(iris.wnoise.umap)) %>% 
  mutate(Species = "predicted")



ggplot(bind_rows(iris.umap_df, iris.wnoise_umap_df), aes(x = V1, y = V2, color = Species)) + geom_point()


```













